opengever/maintenance/scripts/fix_broken_intid_references.py

```diff
@@ -0,0 +1,169 @@
+"""
+Script to fix broken intid references.
+
+    bin/instance run ./scripts/fix_broken_intid_references.py [-n] --check --fix
+
+or with separate check and fix:
+
+   bin/instance run ./scripts/fix_broken_intid_references.py --check -o ./fix-ids
+   bin/instance run ./scripts/fix_broken_intid_references.py --fix -o ./fix-ids
+
+The --check will create a json file (and a backup of it) with all broken object paths
+within the output directory.
+The --fix will then read these paths and fix each object. It will commit the transaction
+every few items and will update the the json with the remeaing items.
+"""
+from five.intid.intid import moveIntIdSubscriber
+from ftw.upgrade.progresslogger import ProgressLogger
+from opengever.maintenance.debughelpers import setup_app
+from opengever.maintenance.debughelpers import setup_option_parser
+from opengever.maintenance.debughelpers import setup_plone
+from plone import api
+from zope.component import getUtility
+from zope.intid import IIntIds
+import json
+import logging
+import os
+import sys
+import time
+import transaction
+
+logger = logging.getLogger('opengever.maintenance')
+handler = logging.StreamHandler(stream=sys.stdout)
+handler.setFormatter(logging.Formatter('%(levelname)s %(message)s'))
+logging.root.addHandler(handler)
+logging.root.setLevel(logging.INFO)
+
+
+class IntIdReferenceFixer(object):
+
+    def __init__(self, portal, dry_run, working_dir):
+        self.portal = portal
+        self.dry_run = dry_run
+        self.working_dir = working_dir
+        self.catalog = api.portal.get_tool('portal_catalog')
+        self.intid_utility = getUtility(IIntIds)
+        self.objs_to_fix_path = os.path.join(working_dir, 'objs_to_fix.json')
+
+    def check(self):
+        broken = []
+        for brain in ProgressLogger(
+                'Check for broken intid references',
+                self.catalog(),
+                logger=logger):
+            obj = brain.getObject()
+            ref = self.intid_utility.refs.get(self.intid_utility.getId(obj))
+            obj_path = '/'.join(obj.getPhysicalPath())
+            if obj_path != ref.path:
+                broken.append(obj_path)
+                logger.info("Found broken object: {}".format(obj_path))
+
+        if not broken:
+            logger.info("Nothing is broken.")
+        else:
+            logger.info("{} objects are broken.".format(len(broken)))
+
+        logger.info("Write broken paths to: {}".format(self.objs_to_fix_path))
+
+        self.write_broken_paths(broken, self.objs_to_fix_path)
+        self.write_broken_paths(broken, self.objs_to_fix_path + '.backup')
+
+    def fix(self, broken, commit_every=100):
+        logger.info("Load broken paths from: {}".format(objs_to_fix_path))
+        broken = self.load_broken_paths(self.objs_to_fix_path)
+
+        if not broken:
+            logger.info("There is nothing to fix.")
+            return
+
+        remaining = broken
```

Comment:
The remaining / fixed stuff is quite weird.

`remaining` gets assigned here to `broken`, but is never modified in this method. So it's never modified at all.
However it gets passed into `self.commit_fixes()` where it is (luckily) not modified but re-assigned. Why not leave out the `remaining = broken` entirely?

Just work with `broken` in this method, and then the code in `self.commit_fixes()` automatically becomes more readable too (~ `remaining = broken - fixed`)